src/rust/engine/fs/store/src/local.rs

```diff
@@ -4,19 +4,394 @@ use super::{EntryType, ShrinkBehavior};
+
+  async fn store_bytes_batch(
+    &self,
+    items: Vec<(Fingerprint, Bytes)>,
+    initial_lease: bool,
+  ) -> Result<(), String>;
+
+  async fn store(
+    &self,
+    initial_lease: bool,
+    src_is_immutable: bool,
+    expected_digest: Digest,
+    src: PathBuf,
+  ) -> Result<(), String>;
+
+  async fn load_bytes_with<
+    T: Send + 'static,
+    F: FnMut(&[u8]) -> Result<T, String> + Send + Sync + 'static,
+  >(
+    &self,
+    fingerprint: Fingerprint,
+    mut f: F,
+  ) -> Result<Option<T>, String>;
+
+  async fn aged_fingerprints(&self) -> Result<Vec<AgedFingerprint>, String>;
+
+  async fn all_digests(&self) -> Result<Vec<Digest>, String> {
+    let fingerprints = self.aged_fingerprints().await?;
+    Ok(
+      fingerprints
+        .into_iter()
+        .map(|fingerprint| Digest {
+          hash: fingerprint.fingerprint,
+          size_bytes: fingerprint.size_bytes,
+        })
+        .collect(),
+    )
+  }
+}
+
+#[async_trait]
+impl UnderlyingByteStore for ShardedLmdb {
+  async fn exists_batch(
+    &self,
+    fingerprints: Vec<Fingerprint>,
+  ) -> Result<HashSet<Fingerprint>, String> {
+    self.exists_batch(fingerprints).await
+  }
+
+  async fn lease(&self, fingerprint: Fingerprint) -> Result<(), String> {
+    self.lease(fingerprint).await
+  }
+
+  async fn remove(&self, fingerprint: Fingerprint) -> Result<bool, String> {
+    self.remove(fingerprint).await
+  }
+
+  async fn store_bytes_batch(
+    &self,
+    items: Vec<(Fingerprint, Bytes)>,
+    initial_lease: bool,
+  ) -> Result<(), String> {
+    self.store_bytes_batch(items, initial_lease).await
+  }
+  async fn store(
+    &self,
+    initial_lease: bool,
+    src_is_immutable: bool,
+    expected_digest: Digest,
+    src: PathBuf,
+  ) -> Result<(), String> {
+    self
+      .store(
+        initial_lease,
+        src_is_immutable,
+        expected_digest,
+        move || std::fs::File::open(&src),
+      )
+      .await
+  }
+
+  async fn load_bytes_with<
+    T: Send + 'static,
+    F: FnMut(&[u8]) -> Result<T, String> + Send + Sync + 'static,
+  >(
+    &self,
+    fingerprint: Fingerprint,
+    f: F,
+  ) -> Result<Option<T>, String> {
+    self.load_bytes_with(fingerprint, f).await
+  }
+
+  async fn aged_fingerprints(&self) -> Result<Vec<AgedFingerprint>, String> {
+    self.all_fingerprints().await
+  }
+}
+
+// We shard so there isn't a plethora of entries in one single dir.
+#[derive(Debug, Clone)]
+pub(crate) struct ShardedFSDB {
+  root: PathBuf,
+  executor: Executor,
+  lease_time: Duration,
+}
+
+impl ShardedFSDB {
+  pub(crate) fn get_path(&self, fingerprint: Fingerprint) -> PathBuf {
+    let hex = fingerprint.to_hex();
+    self.root.join(hex.get(0..2).unwrap()).join(hex)
+  }
+
+  pub(crate) async fn get_tempfile(
+    &self,
+    fingerprint: Fingerprint,
+  ) -> Result<TempImmutableLargeFile, String> {
+    let path = self.get_path(fingerprint);
+    if !path.parent().unwrap().exists() {
```

Comment:
I'd drop this check. AFAIK, the `create_dir_all` call does the equivalent internally, so it should be redundant.